pandas/plotting/_matplotlib/core.py

```diff
@@ -371,6 +438,44 @@ def result(self):
             else:
                 return self.axes[0]
 
+    def _reconstruct_data_with_by(self, data: DataFrame) -> DataFrame:
+        """
+        Internal function to group data, and reassign multiindex column names onto the
+        result in order to let grouped data be used in _compute_plot_data method.
+
+        Parameters
+        ----------
+        data: Original DataFrame to plot
+
+        Returns
+        -------
+        Output is the reconstructed DataFrame with MultiIndex columns. The first level
+        of MI is unique values of groups, and second level of MI is the columns
+        selected by users.
+
+        Examples
+        --------
+        >>> d = {'h': ['h1', 'h1', 'h2'], 'a': [1, 3, 5], 'b': [3, 4, 6]}
+        >>> df = DataFrame(d)
+        >>> _reconstruct_data_with_by(df)
+           h1      h2
+           a   b   a   b
+        0  1   3   NaN NaN
+        1  3   4   NaN NaN
+        2  NaN NaN 5   6
+        """
+        grouped = data.groupby(self.by)
+
+        data_list = []
+        for key, group in grouped:
+            columns = MultiIndex.from_product([[key], self.columns])
+            sub_group = group[self.columns]
```

Comment:
i think the additional functions might be better in a separate module e.g. _matplotlib/grouped.py ?

Reply:
moved